src/predicate.rs

```diff
@@ -185,6 +309,63 @@ impl<A: Predicate, B: Predicate> Predicate for Descendant<A, B> {
                 node = parent;
             }
         }
-        return false;
+        false
+    }
+}
+
+/// Matches if inner Predicate `B` matches the node and `A` matches the node before this one.
+///
+/// This is equivalent to the `before + after` CSS selector.
+#[derive(Copy, Clone, Debug, PartialEq)]
+pub struct ImmediatelyAfter<A, B>(pub A, pub B);
+
+impl<A: Predicate, B: Predicate> Predicate for ImmediatelyAfter<A, B> {
+    fn matches(&self, node: &Node) -> bool {
+        self.1.matches(node) && node.prev().map_or(false, |prev| self.0.matches(&prev))
+    }
+}
+
+/// Matches if inner Predicate `B` matches the node and `A` matches any node before this one.
+///
+/// This is equivalent to the `before ~ after` CSS selector.
+#[derive(Copy, Clone, Debug, PartialEq)]
+pub struct After<A, B>(pub A, pub B);
+
+impl<A: Predicate, B: Predicate> Predicate for After<A, B> {
+    fn matches(&self, node: &Node) -> bool {
+        if self.1.matches(node) {
+            let mut node = *node;
+            while let Some(prev) = node.prev() {
+                if self.0.matches(&prev) {
+                    return true;
+                }
+                node = prev;
+            }
+        }
+        false
+    }
+}
+
+/// Matches a node which has no children, except comments.
+///
+/// This is equivalent to the `:empty` CSS selector.
+#[derive(Copy, Clone, Debug, PartialEq)]
+pub struct Empty;
+
+impl Predicate for Empty {
+    fn matches(&self, node: &Node) -> bool {
+        node.children().all(|n| n.as_comment().is_none())
```

Comment:
Shouldn't this be `.as_comment().is_some()`? i.e. all the children should _be_ comment nodes for this Node to be empty.

Reply:
Yes, you're correct

Reply:
This can also be written as `.all(|n| n.is(Comment))`.